### PR TITLE
Clear coalesced during disable shard

### PR DESCRIPTION
**Pull request by corps**
1.  Move shard skipping logic into processing
2. When skipping shards, clean up the lowest coalesced item.  As more items get created, the lowest coalesced item moves up in subsequent runs, thus hopefully we keep culling under pressure.  It may be possible to scan even coalesced item and cull it during shard skipped phase, but that could also be costly, and there is not easy way to know if it is valuable to do so.  Take the cheap option.
3. Add tests ensuring no infinite loops or regressions during shard skipping.

### CODE REVIEW

**Comment by corps** [comment]
This also impacts synchronous flushing.  When a shard is disabled, synchronous flush will basically keep the item without processing.  This is not good user behavior, but it does protect the system from cascading issues.

File: src/sentry/models/outbox.py [hunk 1]
```diff
@@ -678,11 +686,14 @@ def drain_shard(
                 if _test_processing_barrier:
                     _test_processing_barrier.wait()
 
-                shard_row.process(is_synchronous_flush=not flush_all)
+                processed = shard_row.process(is_synchronous_flush=not flush_all)
```
Comment (GabeVillalobos) [inline src/sentry/models/outbox.py:689]: Won't the query in `self.process_shard` still result in the same index scan failure we've been seeing? It looks like the `select_for_update` precedes this which I believe is the point of failure 🤔

Comment (corps) [reply src/sentry/models/outbox.py:689]: That's ok because _it already has the lock_ in the current session.  It is re-entrant.

Comment (corps) [reply src/sentry/models/outbox.py:689]: The timeout is related to other concurrent processing of the same shard.

File: src/sentry/models/outbox.py [hunk 2]
```diff
@@ -599,12 +600,19 @@ def process_coalesced(
...
+                delete_ids = [item_id for item_id in batch if item_id < coalesced.id]
                 if not len(delete_ids):
                     break
                 self.objects.filter(id__in=delete_ids).delete()
                 deleted_count += len(delete_ids)
 
+            # Only process the highest id after the others have been batch processed.
+            # It's not guaranteed that the ordering of the batch processing is in order,
+            # meaning that failures during deletion could leave an old, staler outbox
+            # alive.
```
Comment (dashed) [inline src/sentry/models/outbox.py:612]: @corps was this the root cause of the outbox backlog?

**Review by GabeVillalobos — APPROVED** [review]

File: src/sentry/models/outbox.py (hunk 2 above)
Comment (corps) [reply src/sentry/models/outbox.py:612]: We're not entirely sure right now what hte outbox backlog was caused by, but suspect org provisioning and or org auth tokens @GabeVillalobos

Comment (dashed) [reply src/sentry/models/outbox.py:612]: @corps we might need a test case for this specific edge case 😅 doesnt have to be in this PR.

**Pull request merged**